modules/Bio/EnsEMBL/Compara/RunnableDB/GenomicAlignBlock/CalculateBlockStats.pm

```diff
@@ -0,0 +1,142 @@
+=head1 LICENSE
+
+Copyright [1999-2015] Wellcome Trust Sanger Institute and the EMBL-European Bioinformatics Institute
+Copyright [2016-2019] EMBL-European Bioinformatics Institute
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+=head1 NAME
+
+Bio::EnsEMBL::Compara::RunnableDB::GenomicAlignBlock::CalculateBlockStats
+
+=head1 DESCRIPTION
+
+Will generate statistics for a list of alignment blocks. The computation is
+offloaded to Utils::Cigars, and this Runnable only has to sum up the statistics.
+
+Two statistics are computed:
+1. The alignment depth, i.e. the number of sequences aligned to a given sequence,
+2. The pairwise coverage
+They are then dataflown to the branches 2 and 3 (alignment depth) and 4
+(pairwise coverage)
+
+=cut
+
+package Bio::EnsEMBL::Compara::RunnableDB::GenomicAlignBlock::CalculateBlockStats;
+
+use strict;
+use warnings;
+
+use List::Util qw(sum);
+
+use Bio::EnsEMBL::Compara::Utils::Cigars;
+use Bio::EnsEMBL::Compara::Utils::Preloader;
+
+use base ('Bio::EnsEMBL::Compara::RunnableDB::BaseRunnable');
+
+
+sub run {
+    my $self = shift @_;
+
+    $self->param('depth_by_genome', {});
+    $self->param('total_pairwise_coverage', {});
+
+    foreach my $genomic_align_block_id (@{$self->param_required('genomic_align_block_ids')}) {
+        $self->process_one_block($genomic_align_block_id);
+    }
+}
+
+
+sub process_one_block {
+    my $self = shift @_;
+    my $genomic_align_block_id = shift @_;
+
+    # Load from the database
+    my $genomic_align_block = $self->compara_dba->get_GenomicAlignBlockAdaptor->fetch_by_dbID($genomic_align_block_id);
+    my $genomic_aligns      = $genomic_align_block->genomic_align_array();
+
+    Bio::EnsEMBL::Compara::Utils::Preloader::load_all_DnaFrags($self->compara_dba->get_DnaFragAdaptor, $genomic_aligns);
+    $self->disconnect_from_databases;
+
+    my $depth_by_genome = $self->param('depth_by_genome');
+
+    # Convert the cigar strings to arrays beforehand and just once
+    my @all_cigar_arrays;
+    my @all_genome_db_ids;
+    while(my $ga = shift @$genomic_aligns) {
+        my $cigar_array  = Bio::EnsEMBL::Compara::Utils::Cigars::get_cigar_array($ga->cigar_line);
+        my $genome_db_id = $ga->dnafrag->genome_db_id;
+        my $i_length     = sum(map {$_->[1]} grep {$_->[0] eq 'I'} @$cigar_array);
```

Comment:
Why are `I`s being handled here?

Reply:
Right. I will add a comment block to explain this.

It's an optimisation. `I`s represent bits that are not aligned to anything, so their depth is automatically 0 and they don't add anything to the coverage. They are handled (correctly) by `column_iterator` but it's relatively expensive as it has to mimic a column of `D`s and `M`s and then call the function that will compute the depth / coverage. It is faster to count them and splice them out here